wordpress-mobile/WordPress-iOS
# Move `blogByBlogId` out of blog service

Moves blog look-ups out of the `BlogService` object.

## Why?

The current approach to looking up a single `Blog` object either has insufficient context for proper threading, or requires state leaking into the service. It also encourages the creation of many intermediate `BlogService` objects that are strictly used for looking up `Blog` objects. This complicates a refactor to avoid global access to the Core Data Stack, because we'd need to thread the stack to these instances. If we instead 

Consider the following pseudo-code examples:

```swift
// An example why we can't just use a single context in a service
class BlogService
{
  init(context: NSManagedObjectContext) {
    self.context = context
  }

  func syncBlogs() {
    let blogs = self.allBlogs()
    let remote = BlogRemote(blog.wpcomAPIRemote)
    remote.fetchAllBlogs { blogs in
      // If we want to perform writes here on a context other than the main one, we need a 
      // temporary context. For that we need the stack.
    }
  }
}
```

```swift
// This is better, but now the BlogService doesn't know how to reliably determine where it 
// should perform operations

class BlogService
{
  init(stack: CoreDataStack) {
    self.coreDataStack = stack
  }

  func updateBlogName(forSiteID siteID: Int, name: String) {
    let context = coreDataStack.newDerivedContext()

    context.performAndWait { context in
      // How do we tell Core Data which context to use for this lookup?
      // We could say "always fetch on the view context", but that precludes lookups in a background context
      let blog = self.blog(withID: siteID)
      blog.name = name
      self.coreDataStack.saveContext(context)
    }
  }
}
```

```swift
// Instead, let's pass the `context` in which we're performing the lookup into the service:
class BlogService
{
  init(stack: CoreDataStack) {
    self.coreDataStack = stack
  }

  func updateBlogName(forSiteID siteID: Int, name: String) {
    let context = coreDataStack.newDerivedContext()

    context.performAndWait { context in
      // How do we tell Core Data which context to use for this lookup?
      // We could say "always fetch on the view context", but that precludes lookups in a background context
      let blog = self.blog(withID: siteID, in: context)
      blog.name = name
      self.coreDataStack.saveContext(context)
    }
  }
  
  func blog(withID siteID: Int, in context: NSManagedObjectContext) -> Blog? {
    let fetchRequest = NSFetchRequest<Self>(entityName: Blog.entityName())
    fetchRequest.predicate = NSPredicate(format: "blogID == %@", siteID)
    return try? context.fetch(fetchRequest).first
  }
}
```

Let's pause here. We've fixed the background problem, but our API is very strange now. Consider trying to use it externally:

```swift
let service = BlogService(stack: self.coreDataStack)
let blog = service.blog(withID: 4, in: self.coreDataStack.viewContext) /// why do I need to provide a context _and_ a stack?!
```

Taking a step back, let's look at the problem this API is trying to solve – we want to be able to fetch `Blog` objects without needing to write the `NSFetchRequest` by hand each time. We see this in other services:

```swift
class PostService 
{
  // [...snip...]

  func downloadPosts(forSite siteID: Int) {
     let api = BlogService(context: self.coreDataStack.viewContext).blog(withID: siteID).wpcomAPIRemote
     let remote = PostRemote(api)
     // [...snip...]
  }
}
```

and even in helpers

```swift
class StatsInsightsHelper 
{
   func getStatsCountForBlog(withID siteID: Int) -> [Stat] {
     return BlogService(content: ContextManager.shared.mainContext).blog(withID: siteID).stats 
   }
}
```


We can improve all of these examples by moving the fetch logic into a static method on the `Blog` object:


```swift
extension Blog {
     static func lookup(withID id: Int, in context: NSManagedObjectContext) -> Blog? {
       let fetchRequest = NSFetchRequest<Self>(entityName: Blog.entityName())
       fetchRequest.predicate = NSPredicate(format: "blogID == %d", id)
       return try? context.fetch(fetchRequest).first
     }
  }
}
```

And here is how those examples would look with the new lookup API:

```swift
class BlogService
{
  init(stack: CoreDataStack) {
    self.coreDataStack = stack
  }

  func updateBlogName(forSiteID siteID: Int, name: String) {
    let context = coreDataStack.newDerivedContext()

    context.performAndWait { context in
      // How do we tell Core Data which context to use for this lookup?
      // We could say "always fetch on the view context", but that precludes lookups in a background context
      let blog = Blog.lookup(withID: siteID, in: context)
      blog.name = name
      self.coreDataStack.saveContext(context)
    }
  }
}

```

```swift
/// no need to provide a stack or have an intermediate object
let blog = Blog.lookup(withID: 4, in: self.coreDataStack.viewContext) 
```

```swift
class PostService 
{
  // [...snip...]

  func downloadPosts(forSite siteID: Int) {
    let api = Blog.lookup(siteID, in: self.coreDataStack.viewContext).wpcomAPIRemote
    let remote = PostRemote(api)
    // [...snip...]
  }
}
```

```swift
class StatsInsightsHelper 
{
   func getStatsCountForBlog(withID siteID: Int) -> [Stat] {
      return Blog.lookup(siteID, in: ContextManager.shared.mainContext).stats
   }
}

// One benefit here is that this method becomes far easier to clean up – it now feels an awfully
// lot like something that could go in a `StatsInsightsService`.
//
// So let's refactor it into one

class StatsInsightsService
{
  init(stack: CoreDataStack) {
    self.coreDataStack = stack
  }

  func getStatsCountForBlog(withID siteID: Int) -> [Stat] {
    // Service methods should only ever be called from the main thread since they're accessed from the UI layer
    assert(Thread.isMainThread) 
    return Blog.lookup(withID: siteID, in: self.coreDataStack).stats
  }
}
```

Requesting review from @aerych and @frosty, as y'all have been around this code a lot for a long time. If both of you would be willing to 👍 it, I'll take that as a sign that it's a good refactoring 😅. Happy to answer questions and/or discuss more – there's a lot of explanation I left out to try to keep this from becoming a novel.

**To test:**
Ensure test pass. To review this PR, if you go commit-by-commit, it should be very clear that the change is mechanical, with no new functionality (except tests)

**PR submission checklist:**
